cupyx/scipy/ndimage/filters.py

```diff
@@ -0,0 +1,191 @@
+            input
+        output (cupy.ndarray or None): The array in which to place the output.
+        mode (str): The array borders are handled according to the given mode
+            (``'reflect'``, ``'constant'``, ``'nearest'``, ``'mirror'``,
+            ``'wrap'``). Default is ``'reflect'``.
+        cval (scalar): Value to fill past edges of input if mode is
+            ``constant``. Default is ``0.0``.
+        origin (scalar or tuple of scalar): The origin parameter controls the
+            placement of the filter, relative to the center of the current
+            element of the input. Default of 0 is equivalent to
+            ``(0,)*input.ndim``.
+
+    Returns:
+        cupy.ndarray: The result of correlate.
+
+    .. seealso:: :func:`scipy.ndimage.correlate`
+    """
+    return _correlate_or_convolve(input, weights, output, mode, cval, origin,
+                                  False)
+
+
+def convolve(input, weights, output=None, mode='reflect', cval=0.0, origin=0):
+    """Multi-dimensional convolution.
+
+    The array is convolved with the given kernel.
+
+    Args:
+        input (cupy.ndarray): The input array.
+        weights (cupy.ndarray): Array of weights, same number of dimensions as
+            input
+        output (cupy.ndarray or None): The array in which to place the output.
+        mode (str): The array borders are handled according to the given mode
+            (``'reflect'``, ``'constant'``, ``'nearest'``, ``'mirror'``,
+            ``'wrap'``). Default is ``'reflect'``.
+        cval (scalar): Value to fill past edges of input if mode is
+            ``constant``. Default is ``0.0``.
+        origin (scalar or tuple of scalar): The origin parameter controls the
+            placement of the filter, relative to the center of the current
+            element of the input. Default of 0 is equivalent to
+            ``(0,)*input.ndim``.
+
+    Returns:
+        cupy.ndarray: The result of convolution.
+
+    .. seealso:: :func:`scipy.ndimage.convolve`
+    """
+    return _correlate_or_convolve(input, weights, output, mode, cval, origin,
+                                  True)
+
+
+def _get_output(output, input, shape=None):
+    if shape is None:
+        shape = input.shape
+    if isinstance(output, cupy.ndarray):
+        if output.shape != tuple(shape):
+            raise ValueError('output shape is not correct')
+    else:
+        dtype = output
+        if dtype is None:
+            dtype = input.dtype
+        output = cupy.zeros(shape, dtype)
+    return output
+
+
+def _correlate_or_convolve(input, weights, output, mode, cval, origin,
+                           convolution):
+    if input.dtype in (numpy.complex64, numpy.complex128, numpy.complex256):
+        raise TypeError('Complex type not supported.')
+    if not hasattr(origin, '__getitem__'):
+        origin = [origin, ] * input.ndim
+    wshape = [ii for ii in weights.shape if ii > 0]
```

Comment:
Why `if ii > 0`?

Reply:
I was also wondering, but I followed SciPy implementatin (https://github.com/scipy/scipy/blob/e1e44d12637997606b1bcc0c6de232349e11eee0/scipy/ndimage/filters.py#L603)

Reply:
I think it's never be evaluated to `False`?
At least it should be tested. How about adding such test cases?

Reply:
Would it be OK just to change it to `wshape = weights.shape` ?

Reply:
That seems fine.
But just to be sure, could you add a test case for 0-size shapes?

Reply:
I added zero-sized weight support (this was not supported by this PR) and its tests.

Reply:
In addition, some tests for invalid cases were also added.

Reply:
Do you still need `if ii > 0`?

Reply:
Yes, it is still necessary to make this PR so compatible to SciPy implementation.

Reply:
OK, I confirmed some of the tests would fail without it.